edn/src/types.rs

```diff
@@ -44,12 +44,96 @@ pub enum Value {
+    Map(BTreeMap<ValueAndSpan, ValueAndSpan>),
+}
+
+/// Span represents the current offset (start, end) into the input string.
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct Span(pub usize, pub usize);
+
+/// A wrapper type around SpannedValue and Span, representing some EDN Value
+/// and the parsing offset (start, end) in the original EDN string.
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct ValueAndSpan {
+    pub inner: SpannedValue,
+    pub span: Option<Span>,
+}
+
+impl From<SpannedValue> for Value {
+    fn from(src: SpannedValue) -> Value {
+        match src {
+            SpannedValue::Nil => Value::Nil,
+            SpannedValue::Boolean(v) => Value::Boolean(v),
+            SpannedValue::Integer(v) => Value::Integer(v),
+            SpannedValue::BigInteger(v) => Value::BigInteger(v),
+            SpannedValue::Float(v) => Value::Float(v),
+            SpannedValue::Text(v) => Value::Text(v),
+            SpannedValue::PlainSymbol(v) => Value::PlainSymbol(v),
+            SpannedValue::NamespacedSymbol(v) => Value::NamespacedSymbol(v),
+            SpannedValue::Keyword(v) => Value::Keyword(v),
+            SpannedValue::NamespacedKeyword(v) => Value::NamespacedKeyword(v),
+            SpannedValue::Vector(v) => Value::Vector(v.into_iter().map(|x| Value::from(x)).collect()),
+            SpannedValue::List(v) => Value::List(v.into_iter().map(|x| Value::from(x)).collect()),
+            SpannedValue::Set(v) => Value::Set(v.into_iter().map(|x| Value::from(x)).collect()),
+            SpannedValue::Map(v) => Value::Map(v.into_iter().map(|(x, y)| (Value::from(x), Value::from(y))).collect()),
+        }
+    }
+}
+
+impl From<Value> for SpannedValue {
+    fn from(src: Value) -> SpannedValue {
+        match src {
+            Value::Nil => SpannedValue::Nil,
+            Value::Boolean(v) => SpannedValue::Boolean(v),
+            Value::Integer(v) => SpannedValue::Integer(v),
+            Value::BigInteger(v) => SpannedValue::BigInteger(v),
+            Value::Float(v) => SpannedValue::Float(v),
+            Value::Text(v) => SpannedValue::Text(v),
+            Value::PlainSymbol(v) => SpannedValue::PlainSymbol(v),
+            Value::NamespacedSymbol(v) => SpannedValue::NamespacedSymbol(v),
+            Value::Keyword(v) => SpannedValue::Keyword(v),
+            Value::NamespacedKeyword(v) => SpannedValue::NamespacedKeyword(v),
+            Value::Vector(v) => SpannedValue::Vector(v.into_iter().map(|x| x.into()).collect()),
+            Value::List(v) => SpannedValue::List(v.into_iter().map(|x| x.into()).collect()),
+            Value::Set(v) => SpannedValue::Set(v.into_iter().map(|x| x.into()).collect()),
+            Value::Map(v) => SpannedValue::Map(v.into_iter().map(|(x, y)| (x.into(), y.into())).collect()),
+        }
+    }
+}
+
+impl From<ValueAndSpan> for Value {
+    fn from(src: ValueAndSpan) -> Value {
+        Value::from(src.inner)
+    }
+}
+
+impl From<Value> for ValueAndSpan {
```

Comment:
The reason for this was because it seemed useful to have a conversion between SpannedValue and Value, but in order to have that, we'd also need to have something from ValueAndSpan to value, because SpannedValue collections have ValueAndSpan children. Not having trivial conversion here meant that a lot of code (like all the from_foo, from_keyword, from_namespace etc. methods would've had to be duplicated. I've ultimately done it with macros to avoid any trivial conversion from Value to ValueAndSpan.